### ISSUE TITLE
Respecting `CMAKE_ARGS` environment variable is unexpected

### ISSUE THREAD

**Issue by vyasr** [scikit-build/scikit-build#722]
Currently scikit-build supports configuring CMake via [two environment variables, `SKBUILD_CONFIGURE_OPTIONS` and `CMAKE_ARGS`](https://scikit-build.readthedocs.io/en/latest/usage.html#environment-variable-configuration). The former is a variable specific to scikit-build and it is unsurprising that it would be used automatically. However, `CMAKE_ARGS` being respected automatically is surprising because CMake itself will not respect those arguments. As discussed in https://github.com/conda-forge/ctng-compiler-activation-feedstock/issues/77#issuecomment-1145389613, this functionality was added to support cross-compilation in conda-forge. However, conda-forge contains libraries in languages other than Python, and it leads to an inconsistent experience when packages that invoke `cmake` directly have to forward the arguments `cmake ${CMAKE_ARGS}` while `python setup.py` commands will use this automatically when using scikit-build. I understand the desire to provide a simpler experience, but this seems like the kind of magic that bites when you're least expecting it. Moreover, since vanilla setuptools has no reason to care about CMake arguments, projects transitioning to scikit-build will need to consider how to forward CMake arguments anyway, so forcing them to pass these arguments explicitly is not surprising behavior.

**Comment by henryiii** [comment]
The main reason CMAKE_ARGS support was added was to support conda-forge; it's a disaster if Setuptools supports conda-forge out of the box but scikit-build doesn't. The settings in there are required to properly support scikit-build - if you were to remove CMAKE_ARGS, scikit-build wouldn't work on conda-forge properly. Also, most of the other environment variable controls for scikit-build have the same name as the keyword argument form - which is `cmake_args` in scikit-build. I think #721 is the correct direction for a solution (and was in an earlier form of this, but I guess got missed?

**Comment by vyasr** [comment]
>  it's a disaster if Setuptools supports conda-forge out of the box but scikit-build doesn't. The settings in there are required to properly support scikit-build - if you were to remove CMAKE_ARGS, scikit-build wouldn't work on conda-forge properly

I agree, scikit-build must work correctly wherever setuptools does. What confuses me is why the `CMAKE_ARGS` are necessary given that setuptools wouldn't honor them in the first place. What information is passed on conda-forge via `CMAKE_ARGS` that was either automatically inferred or unnecessary for setuptools builds?

> Also, most of the other environment variable controls for scikit-build have the same name as the keyword argument form - which is cmake_args in scikit-build. I think https://github.com/scikit-build/scikit-build/issues/721 is the correct direction for a solution (and was in an earlier form of this, but I guess got missed?

Sorry I don't understand what you're trying to say here. It sounds to me like saying that we should be able to pass `cmake_args` as a parameter to `setup(...)` to override the environment variable, but I suspect I'm misinterpreting you given that it's already possible to pass CMake arguments separately from the CLI after the first `--`.

**Comment by henryiii** [comment]
> What information is passed on conda-forge via CMAKE_ARGS that was either automatically inferred or unnecessary for setuptools builds

I don't remember exactly, see some of the links on https://github.com/scikit-build/scikit-build/issues/588. I think it's due to conda having built in support for setuptools specifics (like compiler, linker, cross platform internal variables, etc), and built-in support for CMake, but not built-in support for scikit-build.

`setup(cmake_args=...)` is supported, and CMAKE_ARGS corresponds to that.

**Comment by vyasr** [comment]
@jakirkham it looks like #588 was where the original decision for scikit-build to automatically support the `CMAKE_ARGS` environment variable was made. Based on Henry's [last comment](https://github.com/scikit-build/scikit-build/issues/588#issuecomment-1028063711), do you like that it's automatically respecting `CMAKE_ARGS`, or would you prefer that it used the scikit-build specific `SKBUILD_CONFIGURE_OPTIONS` and packages had to support that? Regarding Henry's concerns in that comment, is it even feasible to do that, or are the potential modifications of `CMAKE_ARGS` such that there really is no way to inject `export SKBUILD_CONFIGURE_OPTIONS="${CMAKE_ARGS}"` at a single point and always get the correct behavior? You know conda much better than I do and would have a better idea of what's possible and what would be the least surprising to package maintainers in the future.

**Comment by henryiii** [comment]
scikit-build (classic) after #1185 is now a thin wrapper around scikit-build-core's setuptools plugin. Closing all open issues; if this still applies, please reopen or open an issue on scikit-build-core. Thanks!
